### DISCUSSION TITLE
Issues using lists

### DISCUSSION THREAD

**Discussion by martirm** [EvgSkv/logica#397] [Q&A] [answered]
Hello,
I am struggling a bit with the use of lists and applying functions to them.  This is a simplified example of what I'm trying to do:

```
HasAttribute(object_id: "1", attribute_id: "01", attribute_name: "A");
HasAttribute(object_id: "1", attribute_id: "02", attribute_name: "B");
HasAttribute(object_id: "2", attribute_id: "01", attribute_name: "A");

Object(object_id: "1", object_name: "first_object");
Object(object_id: "2", object_name: "second_object");
Object(object_id: "3", object_name: "third_object");

HasAttribute_f(name, object_id:) = attribute_id :-
    HasAttribute(object_id:, attribute_id:, attribute_name: name);

HasAttributesAnd(name_list, object_id:, attribute_id_list:) :-
    attribute_id_list List= (HasAttribute_f(name, object_id:) :- name in name_list);

Query(object_id:, object_name:, attribute_id_list:) :-
    HasAttributesAnd(name_list, object_id:, attribute_id_list:),
    Object(object_id:, object_name:),
    Size(attribute_id_list) == Size(name_list),
    name_list == ["A", "B"];
``` 

Basically I have a table of objects and a table of attributes associated with those objects. And given a list of any number of attributes, I want to return the objects that include all of them. Currently this works, but only because I forced  the sizes of **attribute_id_list** and **name_list** to match, (otherwise I would just get all objects). This solution, however, is very slow on large tables, and it seems like there should be a more elegant and efficient way to do it.

Thank you.

**Comment by EvgSkv** [comment, marked as answer]
Hello @martirm !
This is an interesting question.
As I started implementing it I hit duckdb bug and filed an [issue](https://github.com/duckdb/duckdb/issues/14540).
So I'm adjusting the query to avoid the bug.

Generally unions are easier for machines compute than overlaps, so if performance is a concern, then trying to express logic via unions may be a way to go.

Here is my shot at this problem with the goal to keep complexity under control:
https://colab.research.google.com/drive/1DaHmaC1yu6LaRuYrUtwH7fQzHOvmcV9X?usp=sharing

Would you like to try applying it to your data and let me know if it's performant enough?

Thank you!
Evgeny.

**Reply by martirm** [reply]
Thank you, @EvgSkv ! I tried your solution on my data and there is a huge improvement, the queries are now running quite fast. I still need to test it more thoroughly on other tables that follow a similar structure, but it looks like this is the way to go.

**Comment by nikiibayat** [comment]
Hello @EvgSkv 
Thank you for your solution. I have a follow up question. In this example, I'd like to apply a regex so all objects with attribute names similar to user query are returned. Something like: 

ObjectAttributes(object_id:, query_request:) List= attribute_name :- 
  HasAttribute(object_id:, attribute_name:),
  QueryRequest(query_request:),
  true == SqlExpr("attribute_name ~* {query_request}", {query_request:});

Doesn't work as intended. Or instead of using SqlExpr, do you recommend another solution?
db = Postgres

**Reply by EvgSkv** [reply]
Hello @nikiibayat ,
your use of `SqlExpr` looks good to me.
I am looking at your program, and I don't see where `attribute_name` is expected to come from.
Would it be possible to create a colab that would re-produce the issue, so I can simply edit it and make it work? :-)
Otherwise, I am not sure about the context and there are moving pieces which I am not sure how to resolve.
Thank you!
Evgeny.

**Reply by nikiibayat** [reply]
Thank you for your reply @EvgSkv 
I am having issues with comparing list vs string and when to do the SqlExpr
This is the notebook [colab](https://colab.research.google.com/drive/1fhgwnWqXtBqpVvqPFaENjDbP9vei3i_H?usp=sharing)

**Reply by EvgSkv** [reply]
Thanks for sharing @nikiibayat .
Now I see that indeed the error comes because we are trying to match attribute name to a query request, which is a list.
But what is the intention? Should it match if any of the elements match? I created a colab that takes `|` join of the elements of the list:
https://colab.research.google.com/drive/1LWOeKiEKGQS-k8xbDyoJ-8t5k4FTF9Kf?usp=sharing

But I am not sure that this is the logic that you want to implement.

So in your predicate
```
@Ground(ObjectAttributes);
ObjectAttributes(object_id:, query_request:) List= attribute_name :-
  HasAttribute(object_id:, attribute_name:),
  QueryRequest(query_request:),
  re = Join(query_request, "|"),  # <- 
  re ~ Str,                       # <-  These are EvgSkv edits.
  true == SqlExpr("(attribute_name ~* {query_request})", {query_request: re});

```
when do you want `attribute_name` listed? When it matches one of elements of query request, or all? Or something different.

Please let me know when you have a chance :-)

**Reply by nikiibayat** [reply]
Thank you @EvgSkv 
I requested access to the colab you shared. 
In the meantime, I tried your suggested predicate. It doesn't throw errors however, my intention for using ~* is to do a case insensitive search. So if query_request is ["a", "B"] I still expect QueryResult to return o1 which is not the case.

when do you want attribute_name listed? The predicate you suggested is correct. The only issue now is that the SqlExpr seems to not work.

**Reply by EvgSkv** [reply]
> I requested access to the colab you shared.

Ah, sorry, bad link.
This link should work: https://colab.research.google.com/drive/1LWOeKiEKGQS-k8xbDyoJ-8t5k4FTF9Kf?usp=sharing

>  The predicate you suggested is correct. The only issue now is that the SqlExpr seems to not work.

The problem seems to be coming from logic of `QueryResult`

```
QueryResult(query_request:, matching_object:, object_attributes:) :-
  attrs_union = QueryComputationStep2(query_request:, object_id: matching_object, object_attributes:),
  Size(attrs_union) <= Size(object_attributes);
```

There we have the `attrs_union` not matching `object_attributes` if cases change.

So I modified `QueryComputationStep2` as follows:

```
@Ground(QueryComputationStep2);
QueryComputationStep2(query_request:, object_id:, object_attributes:) Set= Lower(x)  :-  # EvgSkv: Added Lower here.
  QueryComputationStep1(query_request:, object_id:, object_attributes:),
  (x in object_attributes | x in query_request);
```

Then it _seems_ to work. Let me know if this resolves the problem or not quite yet.

**Reply by nikiibayat** [reply]
Works like a charm! Thank you @EvgSkv 
Happy Holidays :)

**Reply by EvgSkv** [reply]
Great! Happy holidays to you too @nikiibayat !
